### PR TITLE
[CDAP-13985] UI stack upgrade - React and associated libraries upgrade

### PR DESCRIPTION
**Pull request by ajainarayanan**
JIRA: https://issues.cask.co/browse/CDAP-13985
Build: https://builds.cask.co/browse/CDAP-UDUT23

**TL;DR-**
- Upgrades react and react-dom to 16.4.1
- The purpose is,
    - To be able to use new features (patterns) added as part of the new design.
    - To get perf improvements added as part of the new rewrite in react.
    - To be able to upgrade to newer version of the libraries/components that we use for new features/fixes/perf improvements
- Some new features causes UI to not behave the same way it used to. Fixed most of the cases I saw. This involves fixing
  - keys in list elements
  - passing `undefined` for event handlers instead of null
  - fixing proptypes
- Changed the way we use dlls for dev vs prod. Now in dev build we use dev dll files to be able debug or get stack trace when error happens during development (mainly keys and proptypes).

**Longer version**
- React 15.x -> 16.x was a complete rewrite of react
  - This involves perf improvements on existing functionality
  - Newer features that can make life so much better
- New features that we are directly/indirectly using
  - [Portals](https://reactjs.org/docs/portals.html) (modals)
  - [Context](https://reactjs.org/docs/context.html) (react-redux underneath no longer uses the old context api)
  - [Fragments](https://reactjs.org/docs/fragments.html) (We don't use it directly now but we will very soon)
  - [Error boundaries](https://reactjs.org/docs/error-boundaries.html) (Very basic usage. We need to be able leverage this better to surface 500 js errors and 404 redirects)
  - [Create ref](https://reactjs.org/blog/2018/03/29/react-v-16-3.html#createref-api) (We don't use this feature yet but we need to modify our usage of `ref`s to the new way.)

**Breaking changes**
-  **_ErrorBoundaries_**: Any JS error that happen in the render function will cause react to unmount the entire app. The reasoning is its better to unmount the entire app than showing something that is corrupted/unstable state. Today we have `ErrorBoundary` component that will handle this case. But its not intelligent enough to catch all. The important part is `ErrorBoundary` component will ONLY catch errors that happen in `render` functions of the children ([Error Boundary criteria](https://reactjs.org/docs/error-boundaries.html#introducing-error-boundaries)).

- **_Portals_**: Since modals switched to using portals placing the modal component (say AbstractWizard) inside the button tag will make the modal close right away as the "button" is treated as the parent of modal and all click events that happen in modal is eventually surfaced to parent which handles it as a click on the button and closes the modal. So we need to separate the button that opens the modal and the actual modal component into two different tags(divs)

- **_Lifecycle methods deprecation_** [Reference](https://reactjs.org/blog/2018/03/29/react-v-16-3.html#component-lifecycle-changes) We will eventually should do this to be able to use react's [Async rendering](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html)

**NOTE**:
- The major addition in this PR is because of the new dev dll generated to be used during development.
- There are still some css changes required. This is because of bootstrap patch. Bootstrap 4 defined global font and header sizes to be in rems and we don't maintain the same font size between react and angular. Hence the discrepancies. I didn't address all of them now as it required changing the root (global) styles.

### CODE REVIEW

**Comment by tonybach** [comment]
Did a first pass through the UI. In general things look good, I just found some weird issues with the Add Namespace modal.

namespace-add-problems

The key-value pairs not erasing even when we close the modal is a known issue, but there seems to be something wrong with the KeyValuePairs after the changes, at least when used in this wizard.

**Comment by tonybach** [comment]
In React the label for the 'Add entity' modal doesn't seem centered, and in Angular it becomes really small.

modal-header

**Comment by ajainarayanan** [comment]
Noticed. 
screen shot 2018-08-01 at 7 36 44 am

**Comment by ajainarayanan** [comment]
We need to fix all popovers and tooltips used from `reactstrap` to be replaced with inhouse `Popover` component (because of change in tether usage by reactstrap)

**Comment by tonybach** [comment]
Found a couple of other bugs:
1. Clicking on 'Insights' in DataPrep just shows a spinning wheel forever. In the console, I get the error `Cannot read property 'oneOfType' of undefined`.

2. There's a gap between the namespace dropdown and the navbar, which didn't exist before.

image

3. Clicking on or hovering on the (i) icons in this 'Ingest Data' modal doesn't show any tooltip.

image

4. Found a weird issue with the 'Copy to Clipboard' tooltip in the 'Directives' modal in Analytics.

tooltip-on-popover

File: cdap-ui/app/cdap/components/KeyValuePairs/index.js [hunk 1]
```diff
@@ -92,22 +92,26 @@ let KeyValuePairCopy = connect(
...
-    var { keyValues, onKeyValueChange } = props;
+    var { keyValues } = props;
     this.state = {
       pairs: [...keyValues.pairs]
     };
     KeyValueStore.dispatch({
       type: KeyValueStoreActions.onUpdate,
       payload: {pairs: keyValues.pairs}
     });
+    console.log('Inside keyvalue pairs constructor');
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/KeyValuePairs/index.js]: remove

**Comment by tonybach** [comment]
Need to fix this, otherwise the user won't be able to click 'Play' on any program in Overview 

error-on-fastaction-hover

**Comment by ajainarayanan** [comment]
Have addressed all the issues mentioned above. Things not addressed in this PR are,
Copy to clipboard - https://issues.cask.co/browse/CDAP-14066
Weird gap in namespace dropdown - https://issues.cask.co/browse/CDAP-14067

**Comment by ajainarayanan** [comment]
Will open separate PRs for the referenced JIRAs. Merging this for now.
